src/serac/physics/utilities/functional/tests/functional_comparisons_cuda.cu

```diff
@@ -0,0 +1,446 @@
+
+  mfem::ParBilinearForm B(&fespace);
+
+  mfem::ConstantCoefficient a_coef(a);
+  B.AddDomainIntegrator(new mfem::VectorFEMassIntegrator(a_coef));
+
+  mfem::ConstantCoefficient b_coef(b);
+  B.AddDomainIntegrator(new mfem::CurlCurlIntegrator(b_coef));
+  {
+    SERAC_PROFILE_SCOPE(concat("mfem_localAssemble", postfix));
+    B.Assemble(0);
+  }
+  B.Finalize();
+  std::unique_ptr<mfem::HypreParMatrix> J(
+      SERAC_PROFILE_EXPR(concat("mfem_parallelAssemble", postfix), B.ParallelAssemble()));
+
+  mfem::ParLinearForm             f(&fespace);
+  mfem::VectorFunctionCoefficient load_func(dim, [&](const mfem::Vector& coords, mfem::Vector& output) {
+    double x  = coords(0);
+    double y  = coords(1);
+    output    = 0.0;
+    output(0) = 10 * x * y;
+    output(1) = -5 * (x - y) * y;
+  });
+
+  f.AddDomainIntegrator(new mfem::VectorFEDomainLFIntegrator(load_func));
+  {
+    SERAC_PROFILE_SCOPE(concat("mfem_fAssemble", postfix));
+    f.Assemble();
+  }
+  std::unique_ptr<mfem::HypreParVector> F(
+      SERAC_PROFILE_EXPR(concat("mfem_fParallelAssemble", postfix), f.ParallelAssemble()));
+  F->UseDevice(true);
+
+  mfem::ParGridFunction u_global(&fespace);
+  u_global.Randomize();
+
+  mfem::Vector U(fespace.TrueVSize());
+  U.UseDevice(true);
+  u_global.GetTrueDofs(U);
+
+  using test_space  = decltype(test);
+  using trial_space = decltype(trial);
+
+  Functional<test_space(trial_space), gpu_policy> residual(&fespace, &fespace);
+
+  residual.AddDomainIntegral(Dimension<dim>{}, hcurl_qfunction<dim>{}, mesh);
+
+  // Compute the residual using standard MFEM methods
+  mfem::Vector r1(U.Size());
+  {
+    SERAC_PROFILE_SCOPE(concat("mfem_Apply", postfix));
+    J->Mult(U, r1);
+    r1 -= *F;
+  }
+
+  mfem::Vector r2 = SERAC_PROFILE_EXPR(concat("functional_Apply", postfix), residual(U));
+
+  if (verbose) {
+    std::cout << "||r1||: " << r1.Norml2() << std::endl;
+    std::cout << "||r2||: " << r2.Norml2() << std::endl;
+    std::cout << "||r1-r2||/||r1||: " << mfem::Vector(r1 - r2).Norml2() / r1.Norml2() << std::endl;
+  }
+  EXPECT_NEAR(0., mfem::Vector(r1 - r2).Norml2() / r1.Norml2(), 1.e-13);
+
+  mfem::Operator& grad = SERAC_PROFILE_EXPR(concat("functional_GetGradient", postfix), residual.GetGradient(U));
+
+  mfem::Vector g1 = SERAC_PROFILE_EXPR(concat("mfem_ApplyGradient", postfix), (*J) * U);
+  mfem::Vector g2 = SERAC_PROFILE_EXPR(concat("functional_ApplyGradient", postfix), grad * U);
+
+  if (verbose) {
+    std::cout << "||g1||: " << g1.Norml2() << std::endl;
+    std::cout << "||g2||: " << g2.Norml2() << std::endl;
+    std::cout << "||g1-g2||/||g1||: " << mfem::Vector(g1 - g2).Norml2() / g1.Norml2() << std::endl;
+  }
+  EXPECT_NEAR(0., mfem::Vector(g1 - g2).Norml2() / g1.Norml2(), 1.e-13);
+
+  serac::profiling::terminateCaliper();
+}
+
+TEST(thermal, 2D_linear) { functional_test(H1<1>{}, H1<1>{}, Dimension<2>{}); };
+// TEST(thermal, 2D_quadratic) { functional_test(H1<2>{}, H1<2>{}, Dimension<2>{}); }
```

Comment:
I thought we had a work around so these could all be run from this file?

Reply:
Oops. I did not mean to commit, but I was trying to verify that the reproducer is still necessary. There's still a cupointer issue that corresponds to the reproducer in the test directory.